### DISCUSSION TITLE
TensorRT FP16 Problem

### DISCUSSION THREAD

**Discussion by RyougiKukoc** [AmusementClub/vs-mlrt#7] [Q&A] [answered]
Sorry about that I'm not familiar with the details about the NN behind w2x, but when I tried to execute:
```python
from vsmlrt import Waifu2x, Waifu2xModel, Backend
Waifu2x(src_rgbs, 0, 1, model=Waifu2xModel.cunet, backend=Backend.TRT(fp16=True, num_streams=2)).set_output()
```
The console of trtexec continues logging
> [10/31/2022-20:04:54] [W] [TRT] Weights [name=node_of_Convolution1.weight] had the following issues when converted to FP16:
> [10/31/2022-20:04:54] [W] [TRT]  - Subnormal FP16 values detected.
> [10/31/2022-20:04:54] [W] [TRT] If this is not the desired behavior, please modify the weights or retrain with regularization to reduce the magnitude of the weights.

Shall anybody explain the "issues when converted to FP16"? Is it avoidable?

Machine: 5950x, 3090ti
Environment: VapourSynth R60, vs-mlrt v10

p.s. Comparing fp16=True and fp16=False, the 3x running speed difference shows that it may actually work.

**Comment by WolframRhodium** [comment, marked as answer]
These NNs are usually not trained with quantization in mind: since not all fp32 values can be represented in fp16, the output of fp16 acceleration may differs from fp32 inference. This warning is introduced in TensorRT 8.4 to warn against naive fp16 quantization, but in practice I think it can be ignored in most times.

**Reply by RyougiKukoc** [reply]
Got it, thanks.
